core/src/main/scala/kafka/metrics/clientmetrics/CmClientInformation.scala

```diff
@@ -0,0 +1,78 @@
+package kafka.metrics.clientmetrics
+
+import kafka.Kafka.info
+import kafka.metrics.clientmetrics.CmClientInformation._
+import kafka.network.RequestChannel
+import org.apache.kafka.common.errors.InvalidConfigurationException
+
+
+/**
+ * Information from the client's metadata is gathered from the client's request.
+ */
+object CmClientInformation {
+  val CLIENT_INSTANCE_ID = "client_instance_id"
+  val CLIENT_ID = "client_id"
+  val CLIENT_SOFTWARE_NAME = "client_software_name"
+  val CLIENT_SOFTWARE_VERSION = "client_software_version"
+  val CLIENT_SOURCE_ADDRESS = "client_source_address"
+  val CLIENT_SOURCE_PORT = "client_source_port"
+
+  def apply(request: RequestChannel.Request, clientInstanceId: String): CmClientInformation = {
+    val instance = new CmClientInformation
+    instance.init(clientInstanceId, request.context.clientId(), request.context.clientInformation.softwareName(),
```

Comment:
Is this request context magically available in all requests, or does the client need to specifically add this context somewhere?

Reply:
This information seems to be getting populated from ApiVersionsRequest,  However I let @edenhill or someone who has the expertise confirm that

Reply:
client_software_name and _version requires a previous ApiVersionRequest that all clients send, but it is, I guess, technically possible for a client to skip the ApiVersionRequest and go right for the goodies (PushTelemetryRequest). So a null check may be in place.